backendutil/transform_test.go

```diff
@@ -0,0 +1,323 @@
+	}
+
+	return
+}
+
+func testServerEhlo(t *testing.T, fn ...serverConfigureFunc) (be *backend, s *smtp.Server, c net.Conn, scanner *bufio.Scanner, caps map[string]bool) {
+	be, s, c, scanner = testServerGreeted(t, fn...)
+
+	io.WriteString(c, "EHLO localhost\r\n")
+
+	scanner.Scan()
+	if scanner.Text() != "250-Hello localhost" {
+		t.Fatal("Invalid EHLO response:", scanner.Text())
+	}
+
+	expectedCaps := []string{"PIPELINING", "8BITMIME"}
+	caps = make(map[string]bool)
+
+	for scanner.Scan() {
+		s := scanner.Text()
+
+		if strings.HasPrefix(s, "250 ") {
+			caps[strings.TrimPrefix(s, "250 ")] = true
+			break
+		} else {
+			if !strings.HasPrefix(s, "250-") {
+				t.Fatal("Invalid capability response:", s)
+			}
+			caps[strings.TrimPrefix(s, "250-")] = true
+		}
+	}
+
+	for _, cap := range expectedCaps {
+		if !caps[cap] {
+			t.Fatal("Missing capability:", cap)
+		}
+	}
+
+	return
+}
+
+func testServerAuthenticated(t *testing.T) (be *backend, s *smtp.Server, c net.Conn, scanner *bufio.Scanner) {
+	be, s, c, scanner, caps := testServerEhlo(t)
+
+	if _, ok := caps["AUTH PLAIN"]; !ok {
+		t.Fatal("AUTH PLAIN capability is missing when auth is enabled")
+	}
+
+	io.WriteString(c, "AUTH PLAIN\r\n")
+	scanner.Scan()
+	if scanner.Text() != "334 " {
+		t.Fatal("Invalid AUTH response:", scanner.Text())
+	}
+
+	io.WriteString(c, "AHVzZXJuYW1lAHBhc3N3b3Jk\r\n")
+	scanner.Scan()
+	if !strings.HasPrefix(scanner.Text(), "235 ") {
+		t.Fatal("Invalid AUTH response:", scanner.Text())
+	}
+
+	return
+}
+
+func TestServer(t *testing.T) {
+	be, s, c, scanner := testServerAuthenticated(t)
+	defer s.Close()
+	defer c.Close()
+
+	io.WriteString(c, "MAIL FROM:<root@nsa.gov>\r\n")
+	scanner.Scan()
+	if !strings.HasPrefix(scanner.Text(), "250 ") {
+		t.Fatal("Invalid MAIL response:", scanner.Text())
+	}
+
+	io.WriteString(c, "RCPT TO:<root@gchq.gov.uk>\r\n")
+	scanner.Scan()
+	if !strings.HasPrefix(scanner.Text(), "250 ") {
+		t.Fatal("Invalid RCPT response:", scanner.Text())
+	}
+
+	io.WriteString(c, "DATA\r\n")
+	scanner.Scan()
+	if !strings.HasPrefix(scanner.Text(), "354 ") {
+		t.Fatal("Invalid DATA response:", scanner.Text())
+	}
+
+	io.WriteString(c, "Hey <3\r\n")
+	io.WriteString(c, ".\r\n")
+	scanner.Scan()
+	if !strings.HasPrefix(scanner.Text(), "250 ") {
+		t.Fatal("Invalid DATA response:", scanner.Text())
+	}
+
+	if len(be.messages) != 1 || len(be.anonmsgs) != 0 {
+		t.Fatal("Invalid number of sent messages:", be.messages, be.anonmsgs)
+	}
+
+	msg := be.messages[0]
+	// base64 of "root@nsa.gov"
+	if msg.From != "cm9vdEBuc2EuZ292" {
+		t.Fatal("Invalid mail sender:", msg.From)
+	}
+	// base64 of "root@gchq.gov.uk"
+	if len(msg.To) != 1 || msg.To[0] != "cm9vdEBnY2hxLmdvdi51aw==" {
+		t.Fatal("Invalid mail recipients:", msg.To)
+	}
+	// base64 of "Hey <3\n" (with actual newline)
+	if string(msg.Data) != "SGV5IDwzCg==" {
+		t.Fatal("Invalid mail data:", string(msg.Data))
+	}
+}
+
+func TestServer_tooLongMessage(t *testing.T) {
```

Comment:
Is this test relevant to TransformBackend functionality?

Reply:
Yes, it is there to check if errors are passed through correctly. I missed that part without that test initially. ;-)

Reply:
Alright.